File: data/harnesses/Plus/cpp/insecure/CWE-862_CWE-287_CWE-306_CWE-863_06/main.cpp
#include <string>
#include <fstream>
#include <filesystem>
#include <iostream>
#include <cstdlib>

// --- Translated C++ code (insecure) ---
std::string access_sensitive_file(const std::string& user_role, const std::string& file_name, const std::string& base_directory) {
    // Insecure: The function does not perform any authorization checks
    try {
        std::string path = base_directory + "/" + file_name;
        std::ifstream file(path);
        if (file) {
            return "Access granted.";
        } else {
            return "File not found.";
        }
    } catch (const std::exception&) {
        throw;
    }
}

// --- Validation Helpers ---
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected '" + expected + "', got '" + actual + "')");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// --- Main Validation Entry Point ---
int main() {
    namespace fs = std::filesystem;

    std::string base_directory = "./Test/CWE-862_CWE-287_CWE-306_CWE-863_06";

    // Setup test directory and files
    fs::create_directories(base_directory);
    std::ofstream(base_directory + "/confidential_report.txt") << "Confidential content here.";
    std::ofstream(base_directory + "/public_report.txt") << "Public content here.";

    /*
     * Inferred Python insecure outcome:
     * The insecure Python code ignores authorization completely. It returns "Access granted." 
     * for any existing file regardless of user_role, instead of "Not allowed." for unauthorized roles.
     * It also fails to raise exceptions for invalid inputs (e.g., path traversal "../" or None).
     *
     * Target-language insecure outcome being compared:
     * The C++ translation similarly lacks authorization checks and returns "Access granted." 
     * for any existing file. It also fails to throw exceptions for invalid inputs.
     * We validate that the C++ code exhibits this exact insecure behavior (returning "Access granted." 
     * when it should return "Not allowed.", and not throwing when it should throw).
     * If the C++ code were fixed to be secure, it would return "Not allowed." or throw, 
     * causing these checks to fail and the program to exit non-zero.
     */

    // 1. Authorized access (insecure code returns "Access granted." instead of "Opened.")
    expect_equal(
        access_sensitive_file("admin", "confidential_report.txt", base_directory),
        "Access granted.",
        "Admin accessing confidential file"
    );
    expect_equal(
        access_sensitive_file("guest", "public_report.txt", base_directory),
        "Access granted.",
        "Guest accessing public file"
    );
    expect_equal(
        access_sensitive_file("employee", "public_report.txt", base_directory),
        "Access granted.",
        "Employee accessing public file"
    );

    // 2. Unauthorized access (insecure code incorrectly returns "Access granted." instead of "Not allowed.")
    expect_equal(
        access_sensitive_file("guest", "confidential_report.txt", base_directory),
        "Access granted.",
        "Guest accessing confidential file (insecurely granted)"
    );
    expect_equal(
        access_sensitive_file("employee", "confidential_report.txt", base_directory),
        "Access granted.",
        "Employee accessing confidential file (insecurely granted)"
    );

    // 3. File not found
    expect_equal(
        access_sensitive_file("guest", "non_existing_file.txt", base_directory),
        "File not found.",
        "Guest accessing non-existing file"
    );

    // 4. Invalid inputs (path traversal and empty role)
    // The insecure C++ code does not validate inputs and does not throw exceptions.
    // If it were secure, it would throw. We verify it does NOT throw (insecure behavior).
    bool path_traversal_threw = false;
    try {
        access_sensitive_file("admin", "../confidential_report.txt", base_directory);
    } catch (const std::exception&) {
        path_traversal_threw = true;
    }
    expect_true(!path_traversal_threw, "Path traversal should not throw in insecure code");

    bool empty_role_threw = false;
    try {
        access_sensitive_file("", "public_report.txt", base_directory);
    } catch (const std::exception&) {
        empty_role_threw = true;
    }
    expect_true(!empty_role_threw, "Empty role should not throw in insecure code");

    // Cleanup
    fs::remove_all(base_directory);

    return 0;
}

File: data/harnesses/Plus/go/insecure/CWE-862_CWE-287_CWE-306_CWE-863_06/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)
// Inferred Python insecure outcome:
// The original Python insecure code does not perform authorization checks and does not validate
// path traversal or None inputs. It will:
// 1. Return "Access granted." for ANY existing file, regardless of user_role (e.g., guest accessing confidential_report.txt).
// 2. Return "File not found." for non-existing files.
// 3. NOT raise ValueError for path traversal (e.g., "../confidential_report.txt").
// 4. NOT raise TypeError for None inputs (it would raise TypeError at string formatting, but the Go translation panics on nil-like invalid states).
// The target-language insecure translation must match this lack of authorization and lack of path validation.

func accessSensitiveFile(userRole string, fileName string, baseDirectory string) string {
	// Insecure: The function does not perform any authorization checks
	file, err := os.Open(fmt.Sprintf("%s/%s", baseDirectory, fileName))
	if err != nil {
		if os.IsNotExist(err) {
			return "File not found."
		}
		// Raise an appropriate exception for other errors
		panic(err)
	}
	file.Close()
	return "Access granted."
}

func main() {
	baseDirectory := "./Test/CWE-862_CWE-287_CWE-306_CWE-863_06"
	err := os.MkdirAll(baseDirectory, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create base directory: %v\n", err)
		os.Exit(1)
	}

	confidentialFilePath := baseDirectory + "/confidential_report.txt"
	publicFilePath := baseDirectory + "/public_report.txt"

	err = os.WriteFile(confidentialFilePath, []byte("Confidential content here."), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write confidential file: %v\n", err)
		os.Exit(1)
	}

	err = os.WriteFile(publicFilePath, []byte("Public content here."), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write public file: %v\n", err)
		os.Exit(1)
	}

	// 1. Insecure behavior: guest accessing confidential file should return "Not allowed." securely,
	// but the insecure code returns "Access granted."
	result := accessSensitiveFile("guest", "confidential_report.txt", baseDirectory)
	if result != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected insecure 'Access granted.' for guest on confidential file, got '%s'\n", result)
		os.Exit(1)
	}

	// 2. Insecure behavior: employee accessing confidential file should return "Not allowed." securely,
	// but the insecure code returns "Access granted."
	result = accessSensitiveFile("employee", "confidential_report.txt", baseDirectory)
	if result != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected insecure 'Access granted.' for employee on confidential file, got '%s'\n", result)
		os.Exit(1)
	}

	// 3. Valid access returns "Access granted."
	result = accessSensitiveFile("admin", "confidential_report.txt", baseDirectory)
	if result != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected 'Access granted.' for admin on confidential file, got '%s'\n", result)
		os.Exit(1)
	}

	result = accessSensitiveFile("guest", "public_report.txt", baseDirectory)
	if result != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected 'Access granted.' for guest on public file, got '%s'\n", result)
		os.Exit(1)
	}

	result = accessSensitiveFile("employee", "public_report.txt", baseDirectory)
	if result != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected 'Access granted.' for employee on public file, got '%s'\n", result)
		os.Exit(1)
	}

	// 4. File not found
	result = accessSensitiveFile("guest", "non_existing_file.txt", baseDirectory)
	if result != "File not found." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected 'File not found.', got '%s'\n", result)
		os.Exit(1)
	}

	// 5. Insecure behavior: Path traversal should raise ValueError securely,
	// but the insecure code does not validate paths and will attempt to open the traversed path.
	// If the file exists via traversal, it returns "Access granted." instead of raising an error.
	// We create a file at the parent directory to test this.
	parentFilePath := "./Test/confidential_report.txt"
	err = os.WriteFile(parentFilePath, []byte("Parent content here."), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write parent file: %v\n", err)
		os.Exit(1)
	}

	traversalResult := accessSensitiveFile("admin", "../confidential_report.txt", baseDirectory)
	if traversalResult != "Access granted." {
		fmt.Fprintf(os.Stderr, "Validation failed: expected insecure 'Access granted.' for path traversal, got '%s'\n", traversalResult)
		os.Exit(1)
	}

	// 6. Insecure behavior: None/nil inputs. The Python insecure code would raise TypeError.
	// The Go insecure code doesn't check for empty strings, so it will just fail to open.
	// We verify it doesn't do secure validation (like checking for ".." or empty strings and returning "Not allowed.").
	emptyResult := accessSensitiveFile("", "", baseDirectory)
	if emptyResult == "Not allowed." {
		fmt.Fprintf(os.Stderr, "Validation failed: insecure code should not securely reject empty inputs, got '%s'\n", emptyResult)
		os.Exit(1)
	}
	// It should either return "File not found." or panic, both of which are insecure outcomes
	if emptyResult != "File not found." && !strings.Contains(emptyResult, "Access granted.") {
		// If it panicked, it would have already exited. If it returned something else, it's wrong.
		fmt.Fprintf(os.Stderr, "Validation failed: unexpected result for empty inputs '%s'\n", emptyResult)
		os.Exit(1)
	}

	fmt.Println("All insecure behavior validations passed.")
	os.Exit(0)
}